File: Algorithm/Python/happy_number.py
class Solution:
    # @param {integer} n
    # @return {boolean}
    def isHappy(self, n):
        res = []
        def exract(n):
            lst = []
            while n:
                lst.append(n%10)
                n = n/10
            return lst
        while 1:
            tmp = 0
            
            for i in exract(n):
                tmp += i**2
            if tmp == 1:
                return True
            if tmp in res:
                return False
            res.append(tmp)
            n = tmp

        

if __name__ == '__main__':
    sol = Solution()
    n = 0
    print sol.isHappy(n) 

File: Algorithm/CPP/happy_number.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
using namespace std;


class Solution {
public:
	bool isHappy(int n) {
		vector<int> ans;
		while (true)
		{
			int sum = 0;
			while (n)
			{
				sum = sum+ (n % 10)*(n % 10);
				n = n / 10;
			}
			if (sum == 1) return true;
			
			if (find(ans.begin(), ans.end(), sum) != ans.end()) return false;
			ans.push_back(sum);
			n = sum;
			for (auto k : ans) cout << k << " ";
			cout << endl;
		}

	}

};


int main()
{
	Solution sol;
	vector<int> nums = {1,1,1,0,1
	};
	//cout << sol.numDecodings("101") << endl;
	//cout << sol.numDecodings("100") << endl; 
	//cout << sol.numDecodings("227") << endl;
	cout << sol.isHappy(2) << endl; //1
	//0
	system("pause");
}
